### ISSUE TITLE
volumes_from not running dependent container(?)

### ISSUE THREAD

**Issue by thecatwasnot** [docker/compose#935]
TLDR: volumes_from directive builds but doesn't run the container, container volumes are never actually bound to anything, thus not providing volume data.  Fix is to add a link as well as volumes_from in docker-compose.yml

I have a data container that builds, and as part of build seeds my volume with data, that nginx will then serve.
docker-compose.yml
```yml
 web:
  image: nginx
  volumes_from:
    - blog
  ports:
    - "80:80"
blog:
  build: .
```
And Dockerfile for blog:
```
FROM ruby:2.1-wheezy
MAINTAINER Cole Teeter thecatwasnot@gmail.com

RUN apt-get update \
  && apt-get install apt-transport-https \
  && curl -s https://deb.nodesource.com/gpgkey/nodesource.gpg.key | apt-key add - \
  && echo 'deb https://deb.nodesource.com/node wheezy main' > /etc/apt/sources.list.d/nodesource.list \
  && echo 'deb-src https://deb.nodesource.com/node wheezy main' >> /etc/apt/sources.list.d/nodesource.list \
  && apt-get update \
  && apt-get -y install nodejs

RUN mkdir -p /usr/src/blog \
  && mkdir -p /etc/nginx/sites-enabled/blog \
  && mkdir -p /usr/share/nginx/html
  
COPY docker/nginx-sites.conf /etc/nginx/sites-enabled/blog/
RUN ln -s /usr/src/blog/ /usr/share/nginx/html/blog

WORKDIR /usr/src/blog
COPY Gemfile /usr/src/blog/
COPY Gemfile.lock /usr/src/blog/
RUN bundle install

COPY . /usr/src/blog/

RUN bundle exec jekyll build

VOLUME ["/etc/nginx/sites-enabled/blog","/usr/share/nginx/html/blog"]
CMD ["/bin/true"]
```
Problem is, if I just do a `docker-compose up web` it never runs the blog container, only builds it, and more critically, never binds the volumes it describes to anything.
```
core@core-01 ~ $ docker ps -a
CONTAINER ID        IMAGE               COMMAND                CREATED             STATUS                      PORTS               NAMES
9e7eec9980c1        blog_blog:latest    "/bin/true"            3 minutes ago                                                       blog_blog_1         
1ea00b726ef0        nginx:1             "nginx -g 'daemon of   8 minutes ago       Exited (0) 13 seconds ago                       blog_web_1          
core@core-01 ~ $ docker inspect -f {{.Volumes}} blog_blog_1
<no value>
core@core-01 ~ $ docker start blog_blog_1
blog_blog_1
core@core-01 ~ $ docker ps -a
CONTAINER ID        IMAGE               COMMAND                CREATED             STATUS                          PORTS               NAMES
9e7eec9980c1        blog_blog:latest    "/bin/true"            4 minutes ago       Exited (0) 4 seconds ago                            blog_blog_1         
1ea00b726ef0        nginx:1             "nginx -g 'daemon of   9 minutes ago       Exited (0) About a minute ago                       blog_web_1          
core@core-01 ~ $ docker inspect -f {{.Volumes}} blog_blog_1
map[/etc/nginx/sites-enabled/blog:/var/lib/docker/vfs/dir/76513d43ec6cb15c118165a46778f5e497f98c90ea6012f9c0f98da4ac0e80f6 /usr/share/nginx/html/blog:/var/lib/docker/vfs/dir/59104cf9907d1707f3dd2835ddaf6d8517ecfa592c1ac1e32d58a6d8b47cbaa7]
core@core-01 ~ $ docker -v
Docker version 1.3.3, build 54d900a
```
I'm not sure if this is intended behavior or not.  I expected for the volumes_from directive to basically let me get at the data in the blog service container.  I can get around it by also linking blog to the web service.

**Comment by dnephin** [comment]
What version of docker are you using?

It used to be that it wasn't necessary to start the container, but that may have changed with 1.4.1.

I think #686 would also address this

**Comment by thecatwasnot** [comment]
Docker is 1.3.3

**Comment by Neppord** [comment]
having the same issue with 

```sh
$ docker --version
Docker version 1.6.2, build 7c8fca2
```

**Comment by mbessler** [comment]
I think I'm seeing a similar issue on compose 1.3.0rc1
```
# docker --version
Docker version 1.6.2, build 7c8fca2
# docker-compose --version
docker-compose 1.3.0rc1
```

I see both containers (data volume container and one referencing it via volumes_from) getting launched, but the volumes end up pointing to different locations looking at `docker inspect`:
```
# docker inspect -f {{.Volumes}} syslogdocker_syslogdata_1
map[/var/log:/var/lib/docker/vfs/dir/49606990a03bc9634fac63479f2e7549863ca86fa0c9259e4a03fba7cd03a874]
# docker inspect -f {{.Volumes}} syslogdocker_syslog_1
map[/dev:/tmp/syslogdev /var/log:/var/lib/docker/vfs/dir/d6a77e4c261b991cb72fee1c41f1ccc9d18223299cb7d00d76800150b1843782]
```

On another machine (with `docker-compose 1.2.0` but same docker version) it appears to work as expected.

**Comment by mbessler** [comment]
update: my particular problem seems to be fixed in `docker-compose version: 1.3.0rc2`

**Comment by linuxbender** [comment]
not working with: 
> docker-compose --version
docker-compose version: 1.3.1
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013

 and 
> docker --version
Docker version 1.7.0, build 0baf609

but works with fig :+1:  in my case
> fig --version
fig 1.0.0

**Comment by dnephin** [comment]
A lot has changed since the original issue was opened, and I'm not sure this issue is still relevant. We've made the following changes:
* volumes_from dependencies now start before the container that uses them
* when containers are re-created, we now copy over specific volumes instead of doing a blanket `volumes_from=old_container`.

If you change the contents of a data volume, you must `docker rm` the old container that used that data volume. We always preserve data volumes across restarts (to prevent data loss), so if you change the contents of a data volume, it's up to you to remove it.  I've created #1981 to document and discuss that issue.

#1882 is another issue about changing from data volume to host volume (and not getting what you expect).

Let's continue the triage in those issues.

**Comment by pporada-gl** [comment]
@dnephin 
Could you please post your compose configs and your dockerfiles so that I can see how you're doing this?
* volumes_from dependencies now start before the container that uses them
* when containers are re-created, we now copy over specific volumes instead of doing a blanket volumes_from=old_container.
* If you change the contents of a data volume, you must docker rm the old container that used that data volume.

**Comment by dnephin** [comment]
I'm not sure what you mean. These are changes we made to compose itself. It should apply to everyone using docker-compose 1.4.2.
